### ISSUE TITLE
WAMPCRA fails on python3 because challenge is not json serializable

### ISSUE THREAD

**Issue by jvdm** [crossbario/crossbar#489]
I'm exercising WAMPCRA using crossbar-0.11.1 on python3.

During `HELLO` handling, crossbar is throwing an exception in a
Deferred, the client is never getting the `CHALLENGE` response.

    $ crossbar start
    2015-10-21T19:09:32-0200 [Controller   3442] Automatically choosing optimal Twisted reactor
    2015-10-21T19:09:32-0200 [Controller   3442] Running on Linux and optimal reactor (epoll) was installed.
    2015-10-21T19:09:32-0200 [Controller   3442]      __  __  __  __  __  __      __     __
    2015-10-21T19:09:32-0200 [Controller   3442]     /  `|__)/  \/__`/__`|__) /\ |__)  |/  \
    2015-10-21T19:09:32-0200 [Controller   3442]     \__,|  \\__/.__/.__/|__)/~~\|  \. |\__/
    2015-10-21T19:09:32-0200 [Controller   3442]
    2015-10-21T19:09:32-0200 [Controller   3442]     Version: 0.11.1
    2015-10-21T19:09:32-0200 [Controller   3442]
    2015-10-21T19:09:32-0200 [Controller   3442] Starting from node directory /home/jvdm/.crossbar
    2015-10-21T19:09:32-0200 [Controller   3442] Loading node configuration file '/home/jvdm/.crossbar/config.yaml'
    2015-10-21T19:09:33-0200 [Controller   3442] Entering reactor event loop...
    2015-10-21T19:09:33-0200 [Controller   3442] Warning, could not set process title (setproctitle not installed)
    2015-10-21T19:09:33-0200 [Controller   3442] Process utilities not available
    2015-10-21T19:09:33-0200 [Controller   3442] Joined realm 'crossbar' on node management router
    2015-10-21T19:09:33-0200 [Controller   3442] No WAMPlets detected in enviroment.
    2015-10-21T19:09:33-0200 [Controller   3442] Starting Router with ID 'worker1'...
    2015-10-21T19:09:33-0200 [Router       3446] Worker running under CPython-EPollReactor
    2015-10-21T19:09:33-0200 [Router       3446] Process utilities not available
    2015-10-21T19:09:33-0200 [Controller   3442] Router with ID 'worker1' and PID 3446 started
    2015-10-21T19:09:33-0200 [Controller   3442] Router 'worker1': realm 'realm1' (named realm1) started
    2015-10-21T19:09:33-0200 [Controller   3442] Router 'worker1': role 'role1' (named role1') started on realm 'realm1'
    2015-10-21T19:09:33-0200 [Router       3446] Site starting on 8080
    2015-10-21T19:09:33-0200 [Controller   3442] Router 'worker1': transport 'transport1' started
    ^C2015-10-21T19:09:40-0200 [Controller   3442] Received SIGINT, shutting down.
    2015-10-21T19:09:40-0200 [Controller   3442] sending TERM to subprocess 3446
    2015-10-21T19:09:40-0200 [Controller   3442] waiting for 3446 to exit...
    2015-10-21T19:09:40-0200 [Router       3446] Received SIGTERM, shutting down.
    2015-10-21T19:09:40-0200 [Router       3446] Unhandled error in Deferred:
    2015-10-21T19:09:40-0200 [Router       3446]
    Traceback (most recent call last):
      File "/home/jvdm/venv/lib/python3.4/site-packages/autobahn/wamp/websocket.py", line 96, in onMessage
        self._session.onMessage(msg)
      File "/home/jvdm/venv/lib/python3.4/site-packages/crossbar/router/session.py", line 326, in onMessage
        txaio.add_callbacks(d, success, self._swallow_error_and_abort)
      File "/home/jvdm/venv/lib/python3.4/site-packages/txaio/tx.py", line 386, in add_callbacks
        future.addCallbacks(callback, errback)
      File "/home/jvdm/venv/lib/python3.4/site-packages/twisted/internet/defer.py", line 306, in addCallbacks
        self._runCallbacks()
    --- <exception caught here> ---
      File "/home/jvdm/venv/lib/python3.4/site-packages/twisted/internet/defer.py", line 588, in _runCallbacks
        current.result = callback(current.result, *args, **kw)
      File "/home/jvdm/venv/lib/python3.4/site-packages/crossbar/router/session.py", line 324, in success
        self._transport.send(msg)
      File "/home/jvdm/venv/lib/python3.4/site-packages/autobahn/wamp/websocket.py", line 122, in send
        raise SerializationError("WAMP serialization error ({0})".format(e))
    autobahn.wamp.exception.SerializationError: WAMP serialization error (b'{"nonce": "mvuBd3AH+HDhJB41", "authmethod": "wampcra", "session": 7186254301671638, "authprovider": "static", "authrole": "broker", "authid": "broker", "timestamp": "2015-10-21T21:09:36.517Z"}' is not JSON serializable)

    2015-10-21T19:09:40-0200 [Router       3446] Connection to node controller lost.
    2015-10-21T19:09:40-0200 [Router       3446] No more controller connection; shutting down.
    2015-10-21T19:09:40-0200 [Router       3446] (TCP Port 8080 Closed)
    2015-10-21T19:09:40-0200 [Controller   3442] Process connection gone: A process has ended with a probable error condition: process ended with exit code 1.
    2015-10-21T19:09:40-0200 [Controller   3442] Node worker worker1 ended (0 workers left)
    2015-10-21T19:09:40-0200 [Controller   3442] Node shutting down ..
    2015-10-21T19:09:40-0200 [Controller   3442] Shutting down node...
    2015-10-21T19:09:40-0200 [Controller   3442] Main loop terminated.
    $

I believe this is related to issue #412, now that `WAMP-CRA.challenge`
is `bytes`, you should be careful to `decode()` it to `utf-8` before
serializing in `messages.Challenge`.

In other words:

```diff
diff --git a/crossbar/router/session.py b/crossbar/router/session.py
index 3c2b878..29c769b 100644
--- a/crossbar/router/session.py
+++ b/crossbar/router/session.py
@@ -620,7 +620,7 @@ class RouterSession(_RouterSession):
                                         # send challenge to client
                                         #
                                         extra = {
-                                            u'challenge': self._pending_auth.challenge
+                                            u'challenge': self._pending_auth.challenge.decode('utf-8')
                                         }

                                         # when using salted passwords, provide the client with
```

**Comment by jvdm** [comment]
Actually I've just revisited the issue and IMHO the proper fix is:

```diff
diff --git a/crossbar/router/auth.py b/crossbar/router/auth.py
index e516802..b542fbc 100644
--- a/crossbar/router/auth.py
+++ b/crossbar/router/auth.py
@@ -101,8 +101,8 @@ class PendingAuthWampCra(PendingAuth):
         }

         # challenge must be bytes
-        self.challenge = json.dumps(challenge_obj, ensure_ascii=False).encode('utf8')
-        self.signature = auth.compute_wcs(secret, self.challenge)
+        self.challenge = json.dumps(challenge_obj, ensure_ascii=False)
+        self.signature = auth.compute_wcs(secret, self.challenge.encode('utf8'))


 class PendingAuthTicket(PendingAuth):
```

Since the spec states "it MUST BE a JSON serialized object", I think
it is better to store "challenge" as a utf-8 JSON object and just
encode to bytes to compute the signature.

**Comment by hawkowl** [comment]
Thanks for the bug report.

Could you also please run ``crossbar version``? It doesn't have anything to do with the bug, but I do like to see the Python versions people run CB on ;)

I'll dive a bit deeper, but I think you've hit the nail on the head. I'll have a fix soon.

**Comment by WesleyE** [comment]
With crossbar 0.11.1, I also had to do an .encode on the signature in crossbar/router/session.py @ 853:
`if signature.encode('utf8') == self._pending_auth.signature:`

Otherwise the check would fail, but this might be an additional issue on my end.

```
Automatically choosing optimal Twisted reactor
Running on Linux and optimal reactor (epoll) was installed.
     __  __  __  __  __  __      __     __
    /  `|__)/  \/__`/__`|__) /\ |__)  |/  \
    \__,|  \\__/.__/.__/|__)/~~\|  \. |\__/
                                        
 Crossbar.io        : 0.11.1
   Autobahn         : 0.10.9
     UTF8 Validator : autobahn
     XOR Masker     : autobahn
     JSON Codec     : stdlib
     MsgPack Codec  : -
   Twisted          : 15.4.0-EPollReactor
   Python           : 3.4.0/CPython
 OS                 : Linux-3.19.0-25-generic-x86_64-with-Ubuntu-14.04-trusty
 Machine            : x86_64
```

**Comment by oberstet** [comment]
Works on trunk
